<h1><center>Segmenting and Clustering Neighborhoods in Toronto</center></h1>


____________

<h2><center>Part 1 of 3</center></h2>

**Libraries and Installation**
* **pandas** for data manipulation and analysis
* **requests** for HTTP, definition and usage
* **lxml** for web scraping and XML, HTML files

________________

In [1]:
import pandas as pd
import requests
!pip install lxml

     |████████████████████████████████| 5.5MB 11.0MB/s eta 0:00:01     |█████████████████████▍          | 3.7MB 11.0MB/s eta 0:00:01


In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
link = requests.get(url)
print ('..Loaded')

..Loaded


<h3>1.1 Loading Data for wikipedia link

In [3]:
raw_data = pd.read_html(link.text)
raw_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

<h3>1.2 Dropping unnecessary table and data

In [4]:
raw_data = raw_data[0]
raw_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

<h3>1.3 Dropping "Not assigned" borough

In [5]:
df = raw_data[raw_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

<h3> 1.4 Grouping by Postal Code

In [6]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

<h3> 1.5 Checking for "Not Assigned" Neighbourhood if any so we can change it to borough

In [7]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [8]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

<h3> 1.6 Dropping 'index' and 'level_0' columns

In [9]:
df.drop(['index'], axis = 'columns', inplace = True)

In [10]:
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

<h3> 1.7 Printing number of rows and columns

In [11]:
df.shape

(103, 3)

**There are 103 Rows and 3 Columns**

____________

<h2><center>Part 2 of 3</center></h2>

**Note :** Geocoder package not working as intended, Using the csv data file to load geographical cooradinates

<h3> 2.1 Loading CSV file

In [12]:
gc_data = pd.read_csv("https://cocl.us/Geospatial_data")
gc_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

<h3> 2.2 Checking if both the data sets are equal in dimensions, so we can join the tables

In [13]:
df.shape == gc_data.shape

True

In [14]:
df.dtypes, gc_data.dtypes

(Postal Code      object
 Borough          object
 Neighbourhood    object
 dtype: object,
 Postal Code     object
 Latitude       float64
 Longitude      float64
 dtype: object)

<h3> 2.3 Joining the tables using "Postal code" because it is a common data type

In [15]:
new_data = df.join(gc_data.set_index('Postal Code'), on = 'Postal Code', how = 'inner')
new_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [16]:
new_data.shape

(103, 5)

**There are 103 Rows and 5 Columns**

_______________

<h2><center>Part 3 of 3</center></h2>

**Libraries, Installation and API**
* **numpy** for working with arrays
* **matplotlib.cm** for color map and plotting
* **matplotlib.colors** for colors
* **geopy.geocoders** for loction data
* **folium** for visualizing geospatial data
* **kmeans** from sklearn.cluster for cluster analysis
* **foursquare API** for exploring loactions

In [17]:
!pip install geocoder
!conda install -c conda-forge geopy --yes 
!pip install folium
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as cl
import geocoder
from geopy.geocoders import Nominatim
import folium
from sklearn.cluster import KMeans

     |████████████████████████████████| 102kB 5.5MB/s ta 0:00:011
  Using cached https://files.pythonhosted.org/packages/d2/3d/fa76db83bf75c4f8d338c2fd15c8d33fdd7ad23a9b5e57eb6c5de26b430e/click-7.1.2-py2.py3-none-any.whl
     |████████████████████████████████| 829kB 6.1MB/s eta 0:00:01
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/8b/99/a0/81daf51dcd359a9377b110a8a886b3895921802d2fc1b2397e
Successfully built future
Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         

<h3>3.1 Coordinates for Toronto

In [18]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


<h3>3.2 Visualizing the map with markers of the data

In [20]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)


for latitude, longitude, borough, neighbourhood in zip(new_data['Latitude'], new_data['Longitude'], new_data['Borough'], new_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='darkred',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

<h3>3.3 Setting up Foursquare API for exploring venues

In [21]:
CLIENT_ID = 'CLUN1DBJKMOVKPEJRNKWCTUIOBWOHLIBUHRP0DGJFVAZPXP3' 
CLIENT_SECRET = 'LSKIGR0DZI1055KOZ4CT1DYELCKELESDW2XBYMMMBBEXTYKJ'
VERSION = '20180605'


print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

CLIENT_ID: CLUN1DBJKMOVKPEJRNKWCTUIOBWOHLIBUHRP0DGJFVAZPXP3
CLIENT_SECRET:LSKIGR0DZI1055KOZ4CT1DYELCKELESDW2XBYMMMBBEXTYKJ


<h3> 3.4 Initiating a function to fetch venue categories using API

In [22]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

<h3> 3.5 Fetching venues for every neighbourhood

In [24]:
venues_in_toronto = getNearbyVenues(new_data['Neighbourhood'], new_data['Latitude'], new_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [25]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3             Tim Hortons            Coffee Shop  
4               Portugril  Portuguese Restaurant

<h3> 3.6 Grouping based on neighbourhood

In [26]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   
1310  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572             Tim Hortons   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1306               43.628841               -79.520999        South St. Burger   
1307               43.628841               -79.520999              Wingporium   
1308               43.628841               -79.520999               Dollarama   
1309               43.628841               -79.520999    Artisano Bakery Café   
1310               43.628841               -79.520999          Healthy Planet   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3               Coffee Shop  
4     Portuguese Restaurant  
...                     ...  
1306           Burger Joint  
1307            Wings Joint  
1308         Discount Store  
1309                 Bakery  
1310        Supplement Shop  

[414 rows x 5 columns]

<h3> 3.7 Checking to see how many different categories of venue are there in our data

In [27]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.718518               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[237 rows x 4 columns]

**237 categories of venue**

<h3> 3.8 One hot encoding the categories and adding neighbourhood to complete the encoding

In [28]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1317                  0        0                   0             0   
1318                  0        0                   0             0   
1319                  0        0                   0             0   
1320                  0        0                   0             0   
1321                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1317               0                0                 0                    0   
1318               0                0                 0                    0   
1319               0                0                 0                    0   
1320               0                0                 0                    0   
1321               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  Turkish Restaurant  \
0                0         0  ...              0                   0   
1                0         0  ...              0                   0   
2                0         0  ...              0                   0   
3                0         0  ...              0                   0   
4                0         0  ...              0                   0   
...            ...       ...  ...            ...                 ...   
1317             0         0  ...              0                   0   
1318             0         0  ...              0                   0   
1319             0         0  ...              0                   0   
1320             0         0  ...              0                   0   
1321             0         0  ...              0                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1317                              0                 0                      0   
1318                              0                 0                      0   
1319                              0                 0                      0   
1320                              0                 0                      0   
1321                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4  

In [29]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Turkish Restaurant  \
0                    0             0  ...              0                   0   
1                    0             0  ...              0                   0   
2                    0             0  ...              0                   0   
3                    0             0  ...              0                   0   
4                    0             0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 238 columns]

<h3> 3.9 Calculating mean for categories for each neighbourhood

In [30]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.045455           0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 238 columns]

<h3> 3.10 Function to fetch most common venues

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

**3.10.1** 
* **num_top_venues** value will be 10, to get only the the top 10 most common venues
* **indicators** are Ordinal indicator like "st=first", nd=second, and so on
* **neighbourhood_venues_sorted** will create new dataframe to work with those venues

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1             Sandwich Place              Pharmacy           Coffee Shop   
2                       Bank              Pharmacy    Chinese Restaurant   
3         Chinese Restaurant   Japanese Restaurant                  Café   
4             Sandwich Place    Italian Restaurant       Thai Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store          Dessert Shop     Electronics Store   
1                   Pub                   Gym       Airport Service   
2         Shopping Mall                 Diner        Sandwich Place   
3           Yoga Studio          Dessert Shop     Electronics Store   
4             Juice Bar               Butcher            Restaurant   

         8th Most Common Venue        9th Most Common Venue  \
0  Eastern European Restaurant          Dumpling Restaurant   
1                  Escape Room  Eastern European Restaurant   
2                Deli / Bodega                   Restaurant   
3  Eastern European Restaurant          Dumpling Restaurant   
4                         Café            Indian Restaurant   

  10th Most Common Venue  
0              Drugstore  
1    Dumpling Restaurant  
2            Supermarket  
3              Drugstore  
4       Sushi Restaurant

<h3> 3.11 Using Kmeans for clustering and labelling

**3.11.1**
* **k_num_clusters** number of cluster we will be generating, value will be 5
* **kmeans.lables** for checking the lables of our model

In [33]:
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=0, tol=0.0001, verbose=0)

In [34]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 3, 1, 1, 1, 0, 1, 3, 3, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1,
       3, 1, 3, 1, 1, 1, 3, 1, 1, 1, 3, 1, 1, 1, 4, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1,
       0, 1, 1, 1, 1, 1, 0, 1], dtype=int32)

**3.11.2**
* First adding the **labelled data** to our top 10 venue category data
* Following that **merging** the result to our part 2 data set
* At last dropping the **NaN values** from the data set

In [35]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [37]:
toronto_merged = new_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio      Department Store   
1           Coffee Shop          Intersection          Hockey Arena   
2                  Park                Bakery                   Pub   
3     Accessories Store              Boutique             Gift Shop   
4           Yoga Studio                 Diner        Sandwich Place   

    5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0       Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
1   Portuguese Restaurant            French Restaurant            Comic Shop   
2          Breakfast Spot               Chocolate Shop   Distribution Center   
3  Furniture / Home Store                  Event Space           Coffee Shop   
4   Portuguese Restaurant                         Park    Mexican Restaurant   

     8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0                Drugstore             Donut Shop                Dog Run  
1  Comfort Food Restaurant    Dumpling Restaurant              Drugstore  
2                      Spa           Dessert Shop            Event Space  
3            Women's Store  Vietnamese Restaurant       Airport Terminal  
4       Italian Restaurant             Hobby Shop    Fried Chicken Joint

In [38]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

<h3> 3.12 Using Matplotlib color markers and schemes to plot the clusters on the map

In [39]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# setting color scheme 
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [cl.rgb2hex(i) for i in colors_array]

# adding markers
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

**Note :** there are 5 clusters, light green color might not be visible if not zoomed in.

<h3> 3.13 Cluster Verification

+ **Cluster 1**

In [40]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
68   Central Toronto             0.0         Jewelry Store   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio      Department Store   
21         Women's Store                  Pool           Yoga Studio   
64           Yoga Studio         Deli / Bodega     Electronics Store   
66     Convenience Store           Yoga Studio      Department Store   
68                  Park      Sushi Restaurant                 Trail   
91                 Trail            Playground           Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21                Deli / Bodega  Eastern European Restaurant   
64  Eastern European Restaurant          Dumpling Restaurant   
66            Electronics Store  Eastern European Restaurant   
68                  Yoga Studio                 Dessert Shop   
91             Department Store                 Dessert Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0    Dumpling Restaurant             Drugstore            Donut Shop   
21   Dumpling Restaurant             Drugstore            Donut Shop   
64             Drugstore            Donut Shop               Dog Run   
66   Dumpling Restaurant             Drugstore            Donut Shop   
68    Dim Sum Restaurant                 Diner        Discount Store   
91    Dim Sum Restaurant                 Diner        Discount Store   

   10th Most Common Venue  
0                 Dog Run  
21                Dog Run  
64    Distribution Center  
66                Dog Run  
68    Distribution Center  
91                Dog Run

+ **Cluster 2**

In [41]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0           Pizza Place   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
6         Scarborough             1.0  Fast Food Restaurant   
..                ...             ...                   ...   
97   Downtown Toronto             1.0                  Café   
98          Etobicoke             1.0                  Pool   
99   Downtown Toronto             1.0                   Gym   
100      East Toronto             1.0    Light Rail Station   
102         Etobicoke             1.0                   Gym   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1               Coffee Shop          Intersection          Hockey Arena   
2                      Park                Bakery                   Pub   
3         Accessories Store              Boutique             Gift Shop   
4               Yoga Studio                 Diner        Sandwich Place   
6               Yoga Studio  Ethiopian Restaurant     Electronics Store   
..                      ...                   ...                   ...   
97              Coffee Shop            Restaurant    Seafood Restaurant   
98                    River           Yoga Studio          Dance Studio   
99                 Creperie        Breakfast Spot       Bubble Tea Shop   
100           Auto Workshop            Comic Shop           Pizza Place   
102  Thrift / Vintage Store     Convenience Store        Discount Store   

           5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1          Portuguese Restaurant     French Restaurant            Comic Shop   
2                 Breakfast Spot        Chocolate Shop   Distribution Center   
3         Furniture / Home Store           Event Space           Coffee Shop   
4          Portuguese Restaurant                  Park    Mexican Restaurant   
6    Eastern European Restaurant   Dumpling Restaurant             Drugstore   
..                           ...                   ...                   ...   
97                           Gym              Tea Room             Gastropub   
98   Eastern European Restaurant   Dumpling Restaurant             Drugstore   
99                  Burger Joint    Salon / Barbershop            Restaurant   
100             Recording Studio            Restaurant               Butcher   
102                Burrito Place          Burger Joint        Sandwich Place   

       8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
1    Comfort Food Restaurant    Dumpling Restaurant              Drugstore  
2                        Spa           Dessert Shop            Event Space  
3              Women's Store  Vietnamese Restaurant       Airport Terminal  
4         Italian Restaurant             Hobby Shop    Fried Chicken Joint  
6                 Donut Shop                Dog Run    Distribution Center  
..                       ...                    ...                    ...  
97      Gym / Fitness Center                  Hotel           Concert Hall  
98                Donut Shop                Dog Run    Distribution Center  
99                      Café       Ramen Restaurant                    Pub  
100            Burrito Place             Skate Park                Brewery  
102              Social Club   Fast Food Restaurant                 Bakery  

[84 rows x 12 columns]

+ **Cluster 3**

In [42]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             2.0            Print Shop           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

+ **Cluster 4**

In [43]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10       North York             3.0           Pizza Place   
35        East York             3.0                  Park   
44      Scarborough             3.0              Bus Line   
49       North York             3.0                  Park   
61  Central Toronto             3.0                  Park   
77        Etobicoke             3.0           Pizza Place   
85      Scarborough             3.0            Playground   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10                        Park                   Pub                Bakery   
35           Convenience Store          Intersection           Yoga Studio   
44                      Bakery                  Park          Soccer Field   
49  Construction & Landscaping                Bakery           Yoga Studio   
61                 Swim School              Bus Line           Yoga Studio   
77                        Park              Bus Line        Sandwich Place   
85                        Park                Bakery          Intersection   

   5th Most Common Venue        6th Most Common Venue  \
10   Japanese Restaurant                Deli / Bodega   
35      Department Store            Electronics Store   
44           Bus Station                 Intersection   
49      Department Store            Electronics Store   
61      Department Store  Eastern European Restaurant   
77         Deli / Bodega          Dumpling Restaurant   
85           Yoga Studio             Department Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
10          Dumpling Restaurant             Drugstore            Donut Shop   
35  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
44                Metro Station        Ice Cream Shop       Coworking Space   
49  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
61          Dumpling Restaurant             Drugstore            Donut Shop   
77                    Drugstore            Donut Shop               Dog Run   
85  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

   10th Most Common Venue  
10                Dog Run  
35             Donut Shop  
44               Creperie  
49             Donut Shop  
61                Dog Run  
77    Distribution Center  
85             Donut Shop

+ **Cluster 5**

In [44]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
57   North York             4.0              Baseball Field   
101   Etobicoke             4.0  Construction & Landscaping   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57            Yoga Studio           Event Space           Escape Room   
101        Baseball Field           Yoga Studio          Dessert Shop   

    5th Most Common Venue        6th Most Common Venue  \
57      Electronics Store  Eastern European Restaurant   
101           Escape Room            Electronics Store   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
57           Dumpling Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

    10th Most Common Venue  
57                 Dog Run  
101             Donut Shop

Clusters created sucessfully.

____________________________________
____________________________________